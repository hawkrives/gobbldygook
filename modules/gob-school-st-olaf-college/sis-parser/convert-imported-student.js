// @flow

import {
	Student,
	Schedule,
	type AreaQuery,
	type HydratedStudentType,
	type FabricationType,
} from '@gob/object-student'
import type {Course as CourseType} from '@gob/types'
import flatten from 'lodash/flatten'
import fromPairs from 'lodash/fromPairs'

type PartialCourse = {
	credits: number,
	number: number,
	clbid: string,
	graded: string,
	department: string,
	lab: boolean,
	section: string,
	name: string,

	year: number,
	semester: number,
	term: number,

	_fabrication?: true,
}

type PartialSchedule = {
	semester: number,
	year: number,
	courses: Array<PartialCourse>,
}

export type PartialStudent = {
	courses: Array<PartialCourse>,
	degrees: Array<string>,
	majors: Array<string>,
	concentrations: Array<string>,
	emphases: Array<string>,
	matriculation: number,
	graduation: number,
	advisor: string,
	name: string,
	schedules: Array<PartialSchedule>,
}

type CourseGetter = (
	{clbid: string, term: number},
	?{[key: string]: FabricationType},
) => Promise<
	| CourseType
	| FabricationType
	| {+clbid: string, +term: number, error: string},
>

export async function convertStudent(
	student: PartialStudent,
	getCourse: CourseGetter,
): Promise<HydratedStudentType> {
	let studies = processStudies(student)

	let info
	{
		let {name, advisor, graduation, matriculation} = student
		info = {name, advisor, graduation, matriculation}
	}

	let {schedules, fabrications} = await processSchedules(
		student.schedules,
		getCourse,
	)

	let filledStudent: any = Student({
		...info,
		schedules,
		fabrications,
		studies,
	})

	return filledStudent
}

export async function processSchedules(
	schedules: Array<PartialSchedule>,
	getCourse: CourseGetter,
) {
	let scheds = schedules.map(sched => {
		let clbids = sched.courses.map(c => c.clbid)

		return Schedule({
			...sched,
			clbids,
			active: true,
		})
	})

	let allClbids = flatten(schedules.map(s => s.courses))
	let promisedCourses = allClbids.map(course => {
		let {clbid, term} = course
		return getCourse({clbid, term}).then(resolved => {
			if ('error' in resolved) {
				return {
					...course,
					_fabrication: true,
				}
			}
			return course
		})
	})

	let resolvedCourses = await Promise.all(promisedCourses)

	let fabricationPairs = resolvedCourses
		.filter(c => c._fabrication)
		.map(c => [c.clbid, c])
	let fabrications = fromPairs(fabricationPairs)

	let finalSchedules = fromPairs(scheds.map(s => [s.id, s]))

	return {schedules: finalSchedules, fabrications}
}

export function processStudies({
	majors: m,
	degrees: d,
	emphases: e,
	concentrations: c,
}: PartialStudent): Array<AreaQuery> {
	d = d.map(name => {
		switch (name) {
			case 'B.A.':
				return 'Bachelor of Arts'
			case 'B.M.':
				return 'Bachelor of Music'
			default:
				return name
		}
	})

	return [
		...d.map(name => ({name, type: 'degree', revision: 'latest'})),
		...m.map(name => ({name, type: 'major', revision: 'latest'})),
		...c.map(name => ({name, type: 'concentration', revision: 'latest'})),
		...e.map(name => ({name, type: 'emphasis', revision: 'latest'})),
	]
}
