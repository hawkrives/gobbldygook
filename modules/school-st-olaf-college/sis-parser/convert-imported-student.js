'use strict'
const { Student, Schedule } = require('../../object-student')
const groupBy = require('lodash/groupBy')
const map = require('lodash/map')
const forEach = require('lodash/forEach')
const uniq = require('lodash/uniq')
const fromPairs = require('lodash/fromPairs')
const filter = require('lodash/filter')
const uuid = require('uuid/v4')

module.exports.convertStudent = convertStudent
function convertStudent({ courses, degrees }, getCourse) {
    return Promise.all([
        processSchedules(courses, getCourse),
        processDegrees(degrees),
    ]).then(([schedulesAndFabrications, info]) => {
        let { schedules, fabrications } = schedulesAndFabrications

        return Student(
            Object.assign({}, info, {
                schedules,
                fabrications,
            })
        )
    })
}

module.exports.processSchedules = processSchedules
function processSchedules(courses, getCourse) {
    return Promise.all(
        map(courses, course => {
            return getCourse(course).then(resolvedCourse => {
                if (resolvedCourse.error) {
                    course._fabrication = true
                    course.clbid = course.clbid || uuid()
                    return course
                }
                return resolvedCourse
            })
        })
    ).then(courses => {
        let fabrications = fromPairs(
            map(filter(courses, '_fabrication'), c => [c.clbid, c])
        )

        let schedules = groupBy(courses, 'term')
        schedules = map(schedules, (courses, term) => {
            term = String(term)
            return Schedule({
                courses,
                active: true,
                clbids: map(courses, c => c.clbid),
                year: parseInt(term.substr(0, 4), 10),
                semester: parseInt(term.substr(4, 1), 10),
            })
        })
        schedules = fromPairs(map(schedules, s => [s.id, s]))

        return { schedules, fabrications }
    })
}

module.exports.processDegrees = processDegrees
function processDegrees(degrees) {
    let singularData = resolveSingularDataPoints(degrees)
    let studies = []

    for (let { concentrations, emphases, majors, degree } of degrees) {
        studies.push({ name: degree, type: 'degree', revision: 'latest' })
        studies = studies.concat(
            majors.map(name => ({ name, type: 'major', revision: 'latest' }))
        )
        studies = studies.concat(
            concentrations.map(name => ({
                name,
                type: 'concentration',
                revision: 'latest',
            }))
        )
        studies = studies.concat(
            emphases.map(name => ({
                name,
                type: 'emphasis',
                revision: 'latest',
            }))
        )
    }

    return Object.assign({}, singularData, { studies })
}

module.exports.resolveSingularDataPoints = resolveSingularDataPoints
function resolveSingularDataPoints(degrees) {
    let thereShouldOnlyBeOne = {
        names: map(degrees, d => d.name),
        advisors: map(degrees, d => d.advisor),
        matriculations: map(degrees, d => d.matriculation),
        graduations: map(degrees, d => d.graduation),
    }

    forEach(thereShouldOnlyBeOne, (group, name) => {
        let len = uniq(group).length
        if (len > 1) {
            throw new Error(
                `convertStudent: The student has more than one ${name}: ${JSON.stringify(group)}`
            )
        } else if (!len) {
            throw new Error(
                `convertStudent: The student has no ${name}: ${JSON.stringify(group)}`
            )
        }
    })

    let name = thereShouldOnlyBeOne.names[0]
    let advisor = thereShouldOnlyBeOne.advisors[0]
    let matriculation = parseInt(thereShouldOnlyBeOne.matriculations[0], 10)
    let graduation = parseInt(thereShouldOnlyBeOne.graduations[0], 10)

    return { name, advisor, matriculation, graduation }
}
