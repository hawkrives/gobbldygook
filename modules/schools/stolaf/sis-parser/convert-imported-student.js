import { Student, Schedule } from 'modules/core/student-format'
import groupBy from 'lodash/groupBy'
import map from 'lodash/map'
import forEach from 'lodash/forEach'
import uniq from 'lodash/uniq'
import fromPairs from 'lodash/fromPairs'
import filter from 'lodash/filter'
import uuid from 'uuid/v4'

export function convertStudent({ courses, degrees }, getCourse) {
	return Promise.all([
		processSchedules(courses, getCourse),
		processDegrees(degrees),
	]).then(([ schedulesAndFabrications, info ]) => {
		let { schedules, fabrications } = schedulesAndFabrications

		return Student({
			...info,
			schedules,
			fabrications,
		})
	})
}


export function processSchedules(courses, getCourse) {
	return Promise.all(map(courses, course => {
		return getCourse(course).then(resolvedCourse => {
			if (resolvedCourse.error) {
				course._fabrication = true
				course.clbid = course.clbid || uuid()
				return course
			}
			return resolvedCourse
		})
	})).then(courses => {
		let fabrications = fromPairs(map(filter(courses, '_fabrication'), c => [ c.clbid, c ]))

		let schedules = groupBy(courses, 'term')
		schedules = map(schedules, (courses, term) => {
			term = String(term)
			return Schedule({
				courses,
				active: true,
				clbids: map(courses, c => c.clbid),
				year: parseInt(term.substr(0, 4), 10),
				semester: parseInt(term.substr(4, 1), 10),
			})
		})
		schedules = fromPairs(map(schedules, s => [ s.id, s ]))

		return { schedules, fabrications }
	})
}


export function processDegrees(degrees) {
	let singularData = resolveSingularDataPoints(degrees)
	let studies = []

	for (let { concentrations, emphases, majors, degree } of degrees) {
		studies.push({ name: degree, type: 'degree', revision: 'latest' })
		studies = studies.concat(majors.map(name =>         ({ name, type: 'major', revision: 'latest' })))
		studies = studies.concat(concentrations.map(name => ({ name, type: 'concentration', revision: 'latest' })))
		studies = studies.concat(emphases.map(name =>       ({ name, type: 'emphasis', revision: 'latest' })))
	}

	return {
		...singularData,
		studies,
	}
}


export function resolveSingularDataPoints(degrees) {
	let thereShouldOnlyBeOne = {
		names: map(degrees, d => d.name),
		advisors: map(degrees, d => d.advisor),
		matriculations: map(degrees, d => d.matriculation),
		graduations: map(degrees, d => d.graduation),
	}

	forEach(thereShouldOnlyBeOne, (group, name) => {
		let len = uniq(group).length
		if (len > 1) {
			throw new Error(`convertStudent: The student has more than one ${name}: ${JSON.stringify(group)}`)
		}
		else if (!len) {
			throw new Error(`convertStudent: The student has no ${name}: ${JSON.stringify(group)}`)
		}
	})

	let name = thereShouldOnlyBeOne.names[0]
	let advisor = thereShouldOnlyBeOne.advisors[0]
	let matriculation = parseInt(thereShouldOnlyBeOne.matriculations[0], 10)
	let graduation = parseInt(thereShouldOnlyBeOne.graduations[0], 10)

	return { name, advisor, matriculation, graduation }
}
