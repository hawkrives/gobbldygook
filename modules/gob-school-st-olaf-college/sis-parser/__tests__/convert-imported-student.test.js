import {convertStudent} from '../convert-imported-student'
import {sample} from './__support__/sample-student'
const getCourseMock = async course => course

describe('convertStudent', () => {
	it('converts a student from the imported data to a Gobbldygook student', async () => {
		const actual = await convertStudent(sample, getCourseMock)

		expect(actual).toMatchSnapshot()
	})
})
