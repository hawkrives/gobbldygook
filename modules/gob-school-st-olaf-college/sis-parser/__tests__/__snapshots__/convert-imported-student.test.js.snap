// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertStudent converts a student from the imported data to a Gobbldygook student 1`] = `
Object {
  "advisor": "Hall-Holt, Olaf A.",
  "creditsNeeded": 35,
  "dateCreated": 2018-09-21T21:55:05.698Z,
  "dateLastModified": 2018-09-21T21:55:05.698Z,
  "fabrications": Object {},
  "fulfillments": Object {},
  "graduation": 2016,
  "id": "6c45a725-204e-4431-8e46-9dbd71d90f37",
  "matriculation": 2012,
  "name": "Hawken M. Rives",
  "overrides": Object {},
  "schedules": Object {
    "0849b891-c129-4816-8da8-38c6356878f9": Object {
      "active": true,
      "clbids": Array [
        "0000084513",
        "0000084461",
        "0000085991",
        "0000085992",
        "0000084378",
        "0000084381",
      ],
      "courses": Array [
        Object {
          "clbid": "0000084513",
          "credits": 1,
          "department": "AMCON",
          "graded": "Graded",
          "lab": false,
          "name": "Dem Vistas: 1800-1900",
          "number": 102,
          "section": "B",
          "semester": 3,
          "term": 20123,
          "year": 2012,
        },
        Object {
          "clbid": "0000084461",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "Chinese Calligraphy",
          "number": 268,
          "section": "",
          "semester": 3,
          "term": 20123,
          "year": 2012,
        },
        Object {
          "clbid": "0000085991",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Software Design",
          "number": 251,
          "section": "",
          "semester": 3,
          "term": 20123,
          "year": 2012,
        },
        Object {
          "clbid": "0000085992",
          "credits": 0.25,
          "department": "CSCI",
          "graded": "Graded",
          "lab": true,
          "name": "Software Design/Lab",
          "number": 252,
          "section": "A",
          "semester": 3,
          "term": 20123,
          "year": 2012,
        },
        Object {
          "clbid": "0000084378",
          "credits": 1,
          "department": "JAPAN",
          "graded": "Graded",
          "lab": false,
          "name": "Beginning Japanese II",
          "number": 112,
          "section": "A",
          "semester": 3,
          "term": 20123,
          "year": 2012,
        },
        Object {
          "clbid": "0000084381",
          "credits": 0,
          "department": "JAPAN",
          "graded": "No Grade",
          "lab": true,
          "name": "Beginning Japanese II Lab",
          "number": 112,
          "section": "B",
          "semester": 3,
          "term": 20123,
          "year": 2012,
        },
      ],
      "id": "0849b891-c129-4816-8da8-38c6356878f9",
      "index": 1,
      "metadata": Object {},
      "semester": 3,
      "title": "Schedule G",
      "year": 2012,
    },
    "0c621372-202c-4b80-b8ec-0d15cf847bd9": Object {
      "active": true,
      "clbids": Array [
        "0000089466",
      ],
      "courses": Array [
        Object {
          "clbid": "0000089466",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "AsianCon2: Encounter Asia",
          "number": 215,
          "section": "",
          "semester": 2,
          "term": 20132,
          "year": 2013,
        },
      ],
      "id": "0c621372-202c-4b80-b8ec-0d15cf847bd9",
      "index": 1,
      "metadata": Object {},
      "semester": 2,
      "title": "Schedule L",
      "year": 2013,
    },
    "12d2b974-4615-4c2b-8a7b-4f5bcdee8f33": Object {
      "active": true,
      "clbids": Array [
        "0000089090",
        "0000088273",
        "0000088630",
        "0000088593",
        "0000088681",
        "0000088682",
      ],
      "courses": Array [
        Object {
          "clbid": "0000089090",
          "credits": 1,
          "department": "AMCON",
          "graded": "Graded",
          "lab": false,
          "name": "Remaking Am: 1865-1945",
          "number": 201,
          "section": "A",
          "semester": 1,
          "term": 20131,
          "year": 2013,
        },
        Object {
          "clbid": "0000088273",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "AsianCon1:MappingJourneys",
          "number": 210,
          "section": "",
          "semester": 1,
          "term": 20131,
          "year": 2013,
        },
        Object {
          "clbid": "0000088630",
          "credits": 0.25,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "ID Approaches to Asia",
          "number": 275,
          "section": "",
          "semester": 1,
          "term": 20131,
          "year": 2013,
        },
        Object {
          "clbid": "0000088593",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Hardware Design",
          "number": 241,
          "section": "",
          "semester": 1,
          "term": 20131,
          "year": 2013,
        },
        Object {
          "clbid": "0000088681",
          "credits": 1,
          "department": "JAPAN",
          "graded": "Graded",
          "lab": false,
          "name": "Intermed Japanese I",
          "number": 231,
          "section": "",
          "semester": 1,
          "term": 20131,
          "year": 2013,
        },
        Object {
          "clbid": "0000088682",
          "credits": 0,
          "department": "JAPAN",
          "graded": "No Grade",
          "lab": true,
          "name": "Intermed Japanese I Lab",
          "number": 231,
          "section": "",
          "semester": 1,
          "term": 20131,
          "year": 2013,
        },
      ],
      "id": "12d2b974-4615-4c2b-8a7b-4f5bcdee8f33",
      "index": 1,
      "metadata": Object {},
      "semester": 1,
      "title": "Schedule L",
      "year": 2013,
    },
    "37b29203-9977-4a77-8530-78745ba9ec07": Object {
      "active": true,
      "clbids": Array [
        "0000101017",
        "0000101018",
      ],
      "courses": Array [
        Object {
          "clbid": "0000101017",
          "credits": 1,
          "department": "PHYS",
          "graded": "Graded",
          "lab": false,
          "name": "Musical Acoustics",
          "number": 252,
          "section": "",
          "semester": 2,
          "term": 20152,
          "year": 2015,
        },
        Object {
          "clbid": "0000101018",
          "credits": 0,
          "department": "PHYS",
          "graded": "No Grade",
          "lab": true,
          "name": "Musical Acoustics Lab",
          "number": 252,
          "section": "",
          "semester": 2,
          "term": 20152,
          "year": 2015,
        },
      ],
      "id": "37b29203-9977-4a77-8530-78745ba9ec07",
      "index": 1,
      "metadata": Object {},
      "semester": 2,
      "title": "Schedule 7",
      "year": 2015,
    },
    "60161a2b-68a6-492a-be57-0891a95aeafb": Object {
      "active": true,
      "clbids": Array [
        "0000082908",
        "0000083505",
        "0000082768",
        "0000082771",
        "0000082792",
      ],
      "courses": Array [
        Object {
          "clbid": "0000082908",
          "credits": 1,
          "department": "AMCON",
          "graded": "Graded",
          "lab": false,
          "name": "Declaring Ind: 1607-1865",
          "number": 101,
          "section": "A",
          "semester": 1,
          "term": 20121,
          "year": 2012,
        },
        Object {
          "clbid": "0000083505",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Principles Comp Sci",
          "number": 121,
          "section": "",
          "semester": 1,
          "term": 20121,
          "year": 2012,
        },
        Object {
          "clbid": "0000082768",
          "credits": 1,
          "department": "JAPAN",
          "graded": "Graded",
          "lab": false,
          "name": "Beginning Japanese I",
          "number": 111,
          "section": "A",
          "semester": 1,
          "term": 20121,
          "year": 2012,
        },
        Object {
          "clbid": "0000082771",
          "credits": 0,
          "department": "JAPAN",
          "graded": "No Grade",
          "lab": true,
          "name": "Beginning Japanese I Lab",
          "number": 111,
          "section": "B",
          "semester": 1,
          "term": 20121,
          "year": 2012,
        },
        Object {
          "clbid": "0000082792",
          "credits": 1,
          "department": "REL",
          "graded": "Graded",
          "lab": false,
          "name": "Bible/Culture/Commun",
          "number": 121,
          "section": "C",
          "semester": 1,
          "term": 20121,
          "year": 2012,
        },
      ],
      "id": "60161a2b-68a6-492a-be57-0891a95aeafb",
      "index": 1,
      "metadata": Object {},
      "semester": 1,
      "title": "Schedule D",
      "year": 2012,
    },
    "60c02826-d88b-46c8-9df8-e400a31570c1": Object {
      "active": true,
      "clbids": Array [
        "0000102748",
        "0000100423",
        "0000100346",
        "0000103276",
        "0000099504",
      ],
      "courses": Array [
        Object {
          "clbid": "0000102748",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Operating Systems",
          "number": 273,
          "section": "",
          "semester": 1,
          "term": 20151,
          "year": 2015,
        },
        Object {
          "clbid": "0000100423",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Sem: Senior Capstone",
          "number": 390,
          "section": "A",
          "semester": 1,
          "term": 20151,
          "year": 2015,
        },
        Object {
          "clbid": "0000100346",
          "credits": 0.25,
          "department": "DANCE",
          "graded": "P/N",
          "lab": false,
          "name": "Power Play",
          "number": 115,
          "section": "",
          "semester": 1,
          "term": 20151,
          "year": 2015,
        },
        Object {
          "clbid": "0000103276",
          "credits": 1,
          "department": "HIST",
          "graded": "Graded",
          "lab": false,
          "name": "Women/Medieval Europe",
          "number": 237,
          "section": "",
          "semester": 1,
          "term": 20151,
          "year": 2015,
        },
        Object {
          "clbid": "0000099504",
          "credits": 1,
          "department": "JAPAN",
          "graded": "Audit",
          "lab": false,
          "name": "Advanced Japanese I",
          "number": 301,
          "section": "",
          "semester": 1,
          "term": 20151,
          "year": 2015,
        },
      ],
      "id": "60c02826-d88b-46c8-9df8-e400a31570c1",
      "index": 1,
      "metadata": Object {},
      "semester": 1,
      "title": "Schedule 8",
      "year": 2015,
    },
    "61a115a1-4c59-4329-b6f3-063e734810d4": Object {
      "active": true,
      "clbids": Array [
        "0000085898",
      ],
      "courses": Array [
        Object {
          "clbid": "0000085898",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "Japan SciFi/Global Persp",
          "number": 130,
          "section": "",
          "semester": 2,
          "term": 20122,
          "year": 2012,
        },
      ],
      "id": "61a115a1-4c59-4329-b6f3-063e734810d4",
      "index": 1,
      "metadata": Object {},
      "semester": 2,
      "title": "Schedule 7",
      "year": 2012,
    },
    "6a58314e-652f-4e14-aba8-c901c40e253d": Object {
      "active": true,
      "clbids": Array [
        "0000090719",
        "0000089957",
        "0000090339",
        "0000090340",
        "0000090172",
      ],
      "courses": Array [
        Object {
          "clbid": "0000090719",
          "credits": 1,
          "department": "AMCON",
          "graded": "Graded",
          "lab": false,
          "name": "Pur/Happiness: 1920-Pres.",
          "number": 202,
          "section": "A",
          "semester": 3,
          "term": 20133,
          "year": 2013,
        },
        Object {
          "clbid": "0000089957",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "AsiaCon3:InterpretJourney",
          "number": 220,
          "section": "",
          "semester": 3,
          "term": 20133,
          "year": 2013,
        },
        Object {
          "clbid": "0000090339",
          "credits": 1,
          "department": "JAPAN",
          "graded": "Graded",
          "lab": false,
          "name": "Intermed Japanese II",
          "number": 232,
          "section": "",
          "semester": 3,
          "term": 20133,
          "year": 2013,
        },
        Object {
          "clbid": "0000090340",
          "credits": 0,
          "department": "JAPAN",
          "graded": "No Grade",
          "lab": true,
          "name": "Intermed Japanese II Lab",
          "number": 232,
          "section": "",
          "semester": 3,
          "term": 20133,
          "year": 2013,
        },
        Object {
          "clbid": "0000090172",
          "credits": 1,
          "department": "MATH",
          "graded": "Graded",
          "lab": false,
          "name": "Elem Linear Algebra",
          "number": 220,
          "section": "B",
          "semester": 3,
          "term": 20133,
          "year": 2013,
        },
      ],
      "id": "6a58314e-652f-4e14-aba8-c901c40e253d",
      "index": 1,
      "metadata": Object {},
      "semester": 3,
      "title": "Schedule E",
      "year": 2013,
    },
    "95f59d6b-69c1-4095-9471-93cbc1d060ca": Object {
      "active": true,
      "clbids": Array [
        "0000097582",
        "0000097407",
        "0000095594",
        "0000095842",
        "0000097333",
      ],
      "courses": Array [
        Object {
          "clbid": "0000097582",
          "credits": 1,
          "department": "ART",
          "graded": "Graded",
          "lab": false,
          "name": "World Architecture",
          "number": 161,
          "section": "",
          "semester": 3,
          "term": 20143,
          "year": 2014,
        },
        Object {
          "clbid": "0000097407",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Audit",
          "lab": false,
          "name": "Asian Calligraphy",
          "number": 268,
          "section": "",
          "semester": 3,
          "term": 20143,
          "year": 2014,
        },
        Object {
          "clbid": "0000095594",
          "credits": 1,
          "department": "CSCI",
          "graded": "S/U",
          "lab": false,
          "name": "Algorithms/Data Struc",
          "number": 253,
          "section": "A",
          "semester": 3,
          "term": 20143,
          "year": 2014,
        },
        Object {
          "clbid": "0000095842",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Ethical Iss: Software",
          "number": 263,
          "section": "",
          "semester": 3,
          "term": 20143,
          "year": 2014,
        },
        Object {
          "clbid": "0000097333",
          "credits": 1,
          "department": "REL",
          "graded": "Graded",
          "lab": false,
          "name": "Intro Feminist Theol",
          "number": 209,
          "section": "",
          "semester": 3,
          "term": 20143,
          "year": 2014,
        },
      ],
      "id": "95f59d6b-69c1-4095-9471-93cbc1d060ca",
      "index": 1,
      "metadata": Object {},
      "semester": 3,
      "title": "Schedule W",
      "year": 2014,
    },
    "a88bfa3e-4ef9-4684-83ad-07b5a0ae94f6": Object {
      "active": true,
      "clbids": Array [
        "0000103314",
        "0000103404",
        "0000104968",
        "0000102075",
        "0000103170",
        "0000103157",
      ],
      "courses": Array [
        Object {
          "clbid": "0000103314",
          "credits": 1,
          "department": "AS/PS",
          "graded": "Audit",
          "lab": false,
          "name": "Asian Regionalism",
          "number": 245,
          "section": "",
          "semester": 3,
          "term": 20153,
          "year": 2015,
        },
        Object {
          "clbid": "0000103404",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "Talking in Japan/US",
          "number": 240,
          "section": "",
          "semester": 3,
          "term": 20153,
          "year": 2015,
        },
        Object {
          "clbid": "0000104968",
          "credits": 0.25,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "FLC - Japanese",
          "number": 240,
          "section": "",
          "semester": 3,
          "term": 20153,
          "year": 2015,
        },
        Object {
          "clbid": "0000102075",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "Sem: Asia",
          "number": 399,
          "section": "",
          "semester": 3,
          "term": 20153,
          "year": 2015,
        },
        Object {
          "clbid": "0000103170",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Programming Languages",
          "number": 276,
          "section": "A",
          "semester": 3,
          "term": 20153,
          "year": 2015,
        },
        Object {
          "clbid": "0000103157",
          "credits": 0.25,
          "department": "DANCE",
          "graded": "Graded",
          "lab": false,
          "name": "Contact Improvisation",
          "number": 215,
          "section": "",
          "semester": 3,
          "term": 20153,
          "year": 2015,
        },
      ],
      "id": "a88bfa3e-4ef9-4684-83ad-07b5a0ae94f6",
      "index": 1,
      "metadata": Object {},
      "semester": 3,
      "title": "Schedule 9",
      "year": 2015,
    },
    "b32df035-0659-49b8-8cda-22100cdf0029": Object {
      "active": true,
      "clbids": Array [
        "0000099351",
      ],
      "courses": Array [
        Object {
          "clbid": "0000099351",
          "credits": 1,
          "department": "CHEM",
          "graded": "Graded",
          "lab": false,
          "name": "IR/Software Development",
          "number": 398,
          "section": "B",
          "semester": 2,
          "term": 20142,
          "year": 2014,
        },
      ],
      "id": "b32df035-0659-49b8-8cda-22100cdf0029",
      "index": 1,
      "metadata": Object {},
      "semester": 2,
      "title": "Schedule L",
      "year": 2014,
    },
    "c9a031eb-f893-4b08-804e-55d95d4cd204": Object {
      "active": true,
      "clbids": Array [
        "0000087891",
        "0000081526",
      ],
      "courses": Array [
        Object {
          "clbid": "0000087891",
          "credits": 0,
          "department": "CSCI",
          "graded": "No Grade",
          "lab": false,
          "name": "Intro Pgrm Prob Slvg",
          "number": 1043,
          "section": "P",
          "semester": 9,
          "term": 20119,
          "year": 2011,
        },
        Object {
          "clbid": "0000081526",
          "credits": 1,
          "department": "REG",
          "graded": "No Grade",
          "lab": false,
          "name": "St. Olaf Equiv Credits",
          "number": 1,
          "section": "",
          "semester": 9,
          "term": 20119,
          "year": 2011,
        },
      ],
      "id": "c9a031eb-f893-4b08-804e-55d95d4cd204",
      "index": 1,
      "metadata": Object {},
      "semester": 9,
      "title": "Schedule C",
      "year": 2011,
    },
    "e2e0ffd2-d002-463d-aae4-91e6052915cc": Object {
      "active": true,
      "clbids": Array [
        "0000097217",
        "0000097413",
        "0000097120",
        "0000097119",
        "0000097125",
      ],
      "courses": Array [
        Object {
          "clbid": "0000097217",
          "credits": 1,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "The Philosophy of Anime",
          "number": 230,
          "section": "",
          "semester": 1,
          "term": 20141,
          "year": 2014,
        },
        Object {
          "clbid": "0000097413",
          "credits": 0.25,
          "department": "ASIAN",
          "graded": "Graded",
          "lab": false,
          "name": "FLC - Japanese",
          "number": 230,
          "section": "",
          "semester": 1,
          "term": 20141,
          "year": 2014,
        },
        Object {
          "clbid": "0000097120",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Top: Mobile Web Graphics",
          "number": 300,
          "section": "",
          "semester": 1,
          "term": 20141,
          "year": 2014,
        },
        Object {
          "clbid": "0000097119",
          "credits": 1,
          "department": "CSCI",
          "graded": "Graded",
          "lab": false,
          "name": "Bioinformatics",
          "number": 315,
          "section": "",
          "semester": 1,
          "term": 20141,
          "year": 2014,
        },
        Object {
          "clbid": "0000097125",
          "credits": 1,
          "department": "MATH",
          "graded": "Graded",
          "lab": false,
          "name": "Top: Intro Abstract Math",
          "number": 282,
          "section": "",
          "semester": 1,
          "term": 20141,
          "year": 2014,
        },
      ],
      "id": "e2e0ffd2-d002-463d-aae4-91e6052915cc",
      "index": 1,
      "metadata": Object {},
      "semester": 1,
      "title": "Schedule Z",
      "year": 2014,
    },
  },
  "settings": Object {},
  "studies": Array [
    Object {
      "name": "Bachelor of Arts",
      "revision": "latest",
      "type": "degree",
    },
    Object {
      "name": "Asian Studies",
      "revision": "latest",
      "type": "major",
    },
    Object {
      "name": "Computer Science",
      "revision": "latest",
      "type": "major",
    },
    Object {
      "name": "Japan Studies",
      "revision": "latest",
      "type": "concentration",
    },
  ],
  "version": "3.0.0-test",
}
`;
